### PR TITLE
JENKINS-64403 Declare the page variable "divBasedFormLayout" globally, not only inside forms

### PR DESCRIPTION
**Pull request by EstherAF**
No jira created for this.
Related to [JENKINS-62437](https://issues.jenkins-ci.org/browse/JENKINS-62437).

In some cases the variable `divBasedFormLayout` is not set, because it is only set inside `f:forms`, but not inside standard HTML `form` (which are a lot in jenkins-core and plugins), or `l:ajax`, as some examples.

The goal of this variable is to help developers to make the plugins compatible with the div layout for forms, and despite `form` and `l:ajax` are compatible with the divs layout for forms, the variable is not working on those cases.

In this PR the variable is configured as a global, so it is always set.

### Proposed changelog entries

* Entry 1: The page variable `divBasedFormLayout` is globally available, not only within `<f:form>`

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@fqueiruga @timja @daniel-beck @jsoref @oleg-nenashev 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by jsoref** [comment]
Seems reasonable

**Review by fqueiruga — APPROVED** [review]
LGTM

**Review by timja — APPROVED** [review]
seems reasonable, I assume tested that it's showing up correctly and that it works in a plugin?

**Review by oleg-nenashev — APPROVED** [review]
Do you need this fix in the 2.263.x baseline @EstherAF ? If so, please create a corresponding Jira ticket. It might be reasonable to backport it if it helps developers.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by timja** [comment]
> Do you need this fix in the 2.263.x baseline @EstherAF ? If so, please create a corresponding Jira ticket. It might be reasonable to backport it if it helps developers.

tables to divs isn't getting backported so not needed

**Comment by daniel-beck** [comment]
> I think it is because we are not using the form taglib but the standard `form` HTML tag.

Could you confirm that this is in fact necessary in those plugins? Otherwise it would seem easier to change this to use the standard Jenkins control.

**Comment by EstherAF** [comment]
> seems reasonable, I assume tested that it's showing up correctly and that it works in a plugin?

Yes, it has been tested with private plugin and also with cloudbees-folder-plugin.

> Could you confirm that this is in fact necessary in those plugins? Otherwise, it would seem easier to change this to use the standard Jenkins control.

We don't have a deep understanding of how those plugins are structured, so we want to minimize the changes needed to adapt them to tables-to-divs.

Also, having the variable only in some contexts and not globally is flakier for developers, who would think the variable is available but it isn't, (which is exactly what happened to us) with the risk of not noticing.

**Comment by EstherAF** [comment]
> Do you need this fix in the 2.263.x baseline @EstherAF ? If so, please create a corresponding Jira ticket. It might be reasonable to backport it if it helps developers.

I don't think it is needed, when the variable is not set it will be evaluated to false, which is the right behaviour for 2.263.

**Comment by fqueiruga** [comment]
I don't think it's needed to backport. Please keep in mind that the original variable was introduced in 2.264, so plugins running on <=2.263 will encounter the same behaviour.

**Comment by oleg-nenashev** [comment]
@daniel-beck are you fine with the explanation?

**Comment by daniel-beck** [comment]
-1 from me. I don't think that

> don't have a deep understanding of how those plugins are structured

is a reasonable justification for a core change, or that exploring my recommendation in https://github.com/jenkinsci/jenkins/pull/5075#issuecomment-732423829 first is too much to ask.

(This may be revealing a potential gap in tables-to-div developer docs though, and if so, should be addressed for everyone who hits a similar problem.)

**Comment by timja** [comment]
> (This may be revealing a potential gap in tables-to-div developer docs though, and if so, that should be addressed for everyone who hits a similar problem.)

https://github.com/jenkins-infra/jenkins.io/pull/3990

**Comment by EstherAF** [comment]
I think I may have explained myself poorly. I'm editing the PR to reflect the technical motivation behind this change.

The reason to propose the change is that we shouldn't be forced to use `f:form` to have that variable set. Native HTML form should be supported, because `f:form` is not covering all needs.

**Comment by fqueiruga** [comment]
I think the approach on this PR is the correct one. My reasoning:

- As Esther said, we should support native HTML when possible, not depending on jelly. There can be legitimate uses of `<form>`.
- There may very well be form field partials served as an API. Those would be inside `<l:ajax` but not `<f:form>`.
- Setting it in a variable initialization function is easiest to maintain IMO, rather that have global-ish variable definitions scattered around random jelly widgets. This is not specific to just this PR but more of a general philosophy.

**Comment by daniel-beck** [comment]
> we should support native HTML when possible, not depending on jelly

I expect that extending the support for mixing plain HTML and Jelly tags will make any future changes more cumbersome because people continue to rely on the structure of the emitted HTML. Basically you're enabling developers to continue to do the crazy things that make tables-to-div such a complex change in the first place.

The `f:form` tag is probably not the worst offender here, but it seems confusing to consider this structure supported there and not elsewhere.

**Comment by fqueiruga** [comment]
`<f:form>` is mainly to handle submission, it does nothing for layout. Use of native `<form>` tag does is not an incentive for avoiding using other jelly widgets.

**Comment by alecharp** [comment]
I just tested this when trying to work on [JENKINS-62908](https://issues.jenkins.io/browse/JENKINS-62908) for the ec2 plugin. Without this PR content, the cloud configuration is all messed up, showing multiple time the same configuration. Surely a side effect of https://github.com/jenkinsci/ec2-plugin/blob/master/src/main/resources/lib/ec2/blockWrapper.jelly#L5. But there is no problem when using a Jenkins Core built from this PR.

**Comment by fqueiruga** [comment]
Given that this PR solves issues in more plugins, I think this one should be considered for merge. Maybe for 2.270 even.

**Review by alecharp — APPROVED** [review]

**Comment by fqueiruga** [comment]
What do we do about this one. Does it have enough approvals?

**Comment by fqueiruga** [comment]
@daniel-beck are your concerns a blocker for this one?

**Comment by timja** [comment]
@daniel-beck any thoughts? I'm not sure why it fixes the ec2 plugin issue

cloud configuration uses f.form:
https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/model/GlobalCloudConfiguration/index.groovy#L30

**Comment by amuniz** [comment]
IIUC this is fixing ec2 and we don't know why. If so, I agree with @daniel-beck that we should get to the bottom of it which might lead to a fix somewhere else.

**Comment by fqueiruga** [comment]
The bottom of it is probably that ec2 is not getting it's variable. Maybe that snippet is run inside an f:ajax, for all we know. This is an issue that affects multiple plugins.

Also, we should consider this a tables-to-divs bug. I don't see why there is an issue to accept this when we have merged more obscure and way hackier fixes in JS for some rather ad-hoc failures.

**Review by jtnord — APPROVED** [review]
this solves the issue, esp where things are working outside of `f:form` so I am for this change.

however as stated above the ec2 plugin does appear to be within a `f:form` and as such that seems scary.  without knowing *why* this (and other) plugins do not see the variable scares me, and I am concerned that we have a serious issue that could have other hard to diagnose issues, and I believe the underlying cause should be investigated so that we can completely understand it.

**Comment by amuniz** [comment]
Ah, I missed the `f:ajax` part. Then I think this is a legitimate fix for that case and can be merged as is.

However, we should investigate and get a clear understanding (as part of this PR or not) of why the variable is not visible under some circumstances.

**Review by amuniz — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

**Comment by timja** [comment]
> Ah, I missed the `f:ajax` part. Then I think this is a legitimate fix for that case and can be merged as is.
> 
> However, we should investigate and get a clear understanding (as part of this PR or not) of why the variable is not visible under some circumstances.

we did his this in branch-api and ended up increasing the baseline and dropping compatibility as we couldn't figure it out at the time, this would like fix that too...

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by timja** [comment]
This also resolves https://issues.jenkins.io/browse/JENKINS-64403

**Pull request merged**
